ICE: min_specialization: "When translating substitutions for specialization, the expected specialization failed to hold"

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![crate_type = "lib"]
#![feature(min_specialization)]

#[derive(Copy)]
struct S<const N: [f32; N]>([f32; f32N]);

#[derive(Copy)]
struct S<const N: usize>([f32; N]);

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (f15f0ea73 2023-03-04)
binary: rustc
commit-hash: f15f0ea73972786e426732c5b92ba9a904b866c4
commit-date: 2023-03-04
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 15.0.7
```

### Error output

```
error[E0428]: the name `S` is defined multiple times
 --> b61f5be57812969e152a6c4cfc3e11c8e716c601.rs:8:1
  |
5 | struct S<const N: [f32; N]>([f32; f32N]);
  | ----------------------------------------- previous definition of the type `S` here
...
8 | struct S<const N: usize>([f32; N]);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `S` redefined here
  |
  = note: `S` must be defined only once in the type namespace of this module

error[E0770]: the type of const parameters must not depend on other generic parameters
 --> b61f5be57812969e152a6c4cfc3e11c8e716c601.rs:5:25
  |
5 | struct S<const N: [f32; N]>([f32; f32N]);
  |                         ^ the type must not depend on the parameter `N`

error[E0425]: cannot find value `f32N` in this scope
 --> b61f5be57812969e152a6c4cfc3e11c8e716c601.rs:5:35
  |
5 | struct S<const N: [f32; N]>([f32; f32N]);
  |                                   ^^^^ not found in this scope
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_trait_selection/src/traits/specialize/mod.rs:103:21: When translating substitutions for specialization, the expected specialization failed to hold

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/compiler/rustc_errors/src/lib.rs:1644:9
stack backtrace:
   0:     0x7f1a1fd6731a - std::backtrace_rs::backtrace::libunwind::trace::h4383e0c48d800fd4
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f1a1fd6731a - std::backtrace_rs::backtrace::trace_unsynchronized::h6816d26f08117b4d
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f1a1fd6731a - std::sys_common::backtrace::_print_fmt::hfa640639b16271d1
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f1a1fd6731a - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h4c6c2e6def629543
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f1a1fdcab6e - core::fmt::write::h2c9efad132a334c1
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/core/src/fmt/mod.rs:1232:17
   5:     0x7f1a1fd5a1c5 - std::io::Write::write_fmt::h814bf5ade5b0eb41
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/io/mod.rs:1684:15
   6:     0x7f1a1fd670e5 - std::sys_common::backtrace::_print::h4c645f879d3eaa7a
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7f1a1fd670e5 - std::sys_common::backtrace::print::h3e307a47c9a19510
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7f1a1fd69e5f - std::panicking::default_hook::{{closure}}::h3c9ff47c1a2169d5
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/panicking.rs:271:22
   9:     0x7f1a1fd69b9b - std::panicking::default_hook::hce8f146ffdc07b27
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/panicking.rs:290:9
  10:     0x7f1a22fca2c5 - <rustc_driver_impl[e33d8e87dcf28fa7]::DEFAULT_HOOK::{closure#0}::{closure#0} as core[c90df6780de66b4]::ops::function::FnOnce<(&core[c90df6780de66b4]::panic::panic_info::PanicInfo,)>>::call_once::{shim:vtable#0}
  11:     0x7f1a1fd6a69d - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hbd4f97d1b3fd7f45
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/alloc/src/boxed.rs:2001:9
  12:     0x7f1a1fd6a69d - std::panicking::rust_panic_with_hook::h0f93e1b62c36cd62
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/panicking.rs:696:13
  13:     0x7f1a23526d01 - std[9d05f37f2a0d456e]::panicking::begin_panic::<rustc_errors[30477a6b69c2e257]::ExplicitBug>::{closure#0}
  14:     0x7f1a2351f386 - std[9d05f37f2a0d456e]::sys_common::backtrace::__rust_end_short_backtrace::<std[9d05f37f2a0d456e]::panicking::begin_panic<rustc_errors[30477a6b69c2e257]::ExplicitBug>::{closure#0}, !>
  15:     0x7f1a23594d66 - std[9d05f37f2a0d456e]::panicking::begin_panic::<rustc_errors[30477a6b69c2e257]::ExplicitBug>
  16:     0x7f1a23572c06 - std[9d05f37f2a0d456e]::panic::panic_any::<rustc_errors[30477a6b69c2e257]::ExplicitBug>
  17:     0x7f1a2356e666 - <rustc_errors[30477a6b69c2e257]::HandlerInner>::bug::<&alloc[7c217260d604da75]::string::String>
  18:     0x7f1a2356e330 - <rustc_errors[30477a6b69c2e257]::Handler>::bug::<&alloc[7c217260d604da75]::string::String>
  19:     0x7f1a235654db - rustc_middle[cce008c1292a9385]::util::bug::opt_span_bug_fmt::<rustc_span[9d8165e91f6b103d]::span_encoding::Span>::{closure#0}
  20:     0x7f1a23563e2a - rustc_middle[cce008c1292a9385]::ty::context::tls::with_opt::<rustc_middle[cce008c1292a9385]::util::bug::opt_span_bug_fmt<rustc_span[9d8165e91f6b103d]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f1a23563df6 - rustc_middle[cce008c1292a9385]::ty::context::tls::with_context_opt::<rustc_middle[cce008c1292a9385]::ty::context::tls::with_opt<rustc_middle[cce008c1292a9385]::util::bug::opt_span_bug_fmt<rustc_span[9d8165e91f6b103d]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f1a23565426 - rustc_middle[cce008c1292a9385]::util::bug::opt_span_bug_fmt::<rustc_span[9d8165e91f6b103d]::span_encoding::Span>
  23:     0x7f1a216deb73 - rustc_middle[cce008c1292a9385]::util::bug::bug_fmt
  24:     0x7f1a21605896 - rustc_trait_selection[6c766aae540ef51f]::traits::specialize::translate_substs
  25:     0x7f1a22355b69 - rustc_hir_analysis[e0a40ec4e05edc38]::impl_wf_check::check_mod_impl_wf
  26:     0x7f1a22651841 - rustc_query_system[637feee9637b99c]::query::plumbing::try_execute_query::<rustc_query_impl[2e65c241cbff84cd]::queries::check_mod_impl_wf, rustc_query_impl[2e65c241cbff84cd]::plumbing::QueryCtxt>
  27:     0x7f1a226513ea - <rustc_query_impl[2e65c241cbff84cd]::Queries as rustc_middle[cce008c1292a9385]::ty::query::QueryEngine>::check_mod_impl_wf
  28:     0x7f1a212e871f - <rustc_middle[cce008c1292a9385]::hir::map::Map>::for_each_module::<rustc_hir_analysis[e0a40ec4e05edc38]::check_crate::{closure#2}::{closure#0}::{closure#0}>
  29:     0x7f1a212e8218 - <rustc_session[1facd9ff6612a9d8]::session::Session>::track_errors::<rustc_hir_analysis[e0a40ec4e05edc38]::check_crate::{closure#2}, ()>
  30:     0x7f1a212e5673 - rustc_hir_analysis[e0a40ec4e05edc38]::check_crate
  31:     0x7f1a212d9a22 - rustc_interface[6a01411731a6f2f2]::passes::analysis
  32:     0x7f1a228291fd - rustc_query_system[637feee9637b99c]::query::plumbing::try_execute_query::<rustc_query_impl[2e65c241cbff84cd]::queries::analysis, rustc_query_impl[2e65c241cbff84cd]::plumbing::QueryCtxt>
  33:     0x7f1a22828eff - <rustc_query_impl[2e65c241cbff84cd]::Queries as rustc_middle[cce008c1292a9385]::ty::query::QueryEngine>::analysis
  34:     0x7f1a226539d6 - <rustc_middle[cce008c1292a9385]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[e33d8e87dcf28fa7]::run_compiler::{closure#1}::{closure#2}::{closure#4}, core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>>
  35:     0x7f1a22229238 - rustc_span[9d8165e91f6b103d]::with_source_map::<core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>, rustc_interface[6a01411731a6f2f2]::interface::run_compiler<core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>, rustc_driver_impl[e33d8e87dcf28fa7]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  36:     0x7f1a222208bc - std[9d05f37f2a0d456e]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[6a01411731a6f2f2]::util::run_in_thread_pool_with_globals<rustc_interface[6a01411731a6f2f2]::interface::run_compiler<core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>, rustc_driver_impl[e33d8e87dcf28fa7]::run_compiler::{closure#1}>::{closure#0}, core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>>
  37:     0x7f1a222202ea - <<std[9d05f37f2a0d456e]::thread::Builder>::spawn_unchecked_<rustc_interface[6a01411731a6f2f2]::util::run_in_thread_pool_with_globals<rustc_interface[6a01411731a6f2f2]::interface::run_compiler<core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>, rustc_driver_impl[e33d8e87dcf28fa7]::run_compiler::{closure#1}>::{closure#0}, core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[c90df6780de66b4]::result::Result<(), rustc_span[9d8165e91f6b103d]::ErrorGuaranteed>>::{closure#1} as core[c90df6780de66b4]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7f1a1fd74593 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h437f2034e21e0804
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/alloc/src/boxed.rs:1987:9
  39:     0x7f1a1fd74593 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h9ba15d2eeb7a9119
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/alloc/src/boxed.rs:1987:9
  40:     0x7f1a1fd74593 - std::sys::unix::thread::Thread::new::thread_start::h41ec441a67c2e320
                               at /rustc/f15f0ea73972786e426732c5b92ba9a904b866c4/library/std/src/sys/unix/thread.rs:108:17
  41:     0x7f1a1fb06bb5 - <unknown>
  42:     0x7f1a1fb88d90 - <unknown>
  43:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.70.0-nightly (f15f0ea73 2023-03-04) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [check_mod_impl_wf] checking that impls are well-formed in top-level module
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 4 previous errors

Some errors have detailed explanations: E0425, E0428, E0770.
For more information about an error, try `rustc --explain E0425`.
```

</p>
</details>

--- a/compiler/rustc_infer/src/infer/combine.rs
+++ b/compiler/rustc_infer/src/infer/combine.rs
@@ -34,7 +34,6 @@ use rustc_hir::def_id::DefId;
 use rustc_middle::infer::canonical::OriginalQueryValues;
 use rustc_middle::infer::unify_key::{ConstVarValue, ConstVariableValue};
 use rustc_middle::infer::unify_key::{ConstVariableOrigin, ConstVariableOriginKind};
-use rustc_middle::traits::query::NoSolution;
 use rustc_middle::traits::ObligationCause;
 use rustc_middle::ty::error::{ExpectedFound, TypeError};
 use rustc_middle::ty::relate::{self, Relate, RelateResult, TypeRelation};
@@ -161,9 +160,9 @@ impl<'tcx> InferCtxt<'tcx> {
         //
         // This probe is probably not strictly necessary but it seems better to be safe and not accidentally find
         // ourselves with a check to find bugs being required for code to compile because it made inference progress.
-        self.probe(|_| {
+        let compatible_types = self.probe(|_| {
             if a.ty() == b.ty() {
-                return;
+                return Ok(());
             }
 
             // We don't have access to trait solving machinery in `rustc_infer` so the logic for determining if the
@@ -173,14 +172,23 @@ impl<'tcx> InferCtxt<'tcx> {
                 (relation.param_env(), a.ty(), b.ty()),
                 &mut OriginalQueryValues::default(),
             );
-
-            if let Err(NoSolution) = self.tcx.check_tys_might_be_eq(canonical) {
+            self.tcx.check_tys_might_be_eq(canonical).map_err(|_| {
                 self.tcx.sess.delay_span_bug(
                     DUMMY_SP,
                     &format!("cannot relate consts of different types (a={:?}, b={:?})", a, b,),
-                );
-            }
+                )
+            })
         });
+
+        // If the consts have differing types, just bail with a const error with
+        // the expected const's type. Specifically, we don't want const infer vars
+        // to do any type shapeshifting before and after resolution.
+        if let Err(guar) = compatible_types {
+            return Ok(self.tcx.const_error_with_guaranteed(
+                if relation.a_is_expected() { a.ty() } else { b.ty() },
+                guar,
+            ));
+        }
 
         match (a.kind(), b.kind()) {
             (

--- a/compiler/rustc_middle/src/ty/consts.rs
+++ b/compiler/rustc_middle/src/ty/consts.rs
@@ -135,6 +135,9 @@ impl<'tcx> Const<'tcx> {
                 _,
                 &hir::Path { res: Res::Def(DefKind::ConstParam, def_id), .. },
             )) => {
+                // Use the type from the param's definition, since we can resolve it,
+                // not the expected parameter type from WithOptConstParam.
+                let param_ty = tcx.type_of(def_id).subst_identity();
                 match tcx.named_bound_var(expr.hir_id) {
                     Some(rbv::ResolvedArg::EarlyBound(_)) => {
                         // Find the name and index of the const parameter by indexing the generics of
@@ -143,14 +146,14 @@ impl<'tcx> Const<'tcx> {
                         let generics = tcx.generics_of(item_def_id);
                         let index = generics.param_def_id_to_index[&def_id];
                         let name = tcx.item_name(def_id);
-                        Some(tcx.mk_const(ty::ParamConst::new(index, name), ty))
+                        Some(tcx.mk_const(ty::ParamConst::new(index, name), param_ty))
                     }
                     Some(rbv::ResolvedArg::LateBound(debruijn, index, _)) => Some(tcx.mk_const(
                         ty::ConstKind::Bound(debruijn, ty::BoundVar::from_u32(index)),
-                        ty,
+                        param_ty,
                     )),
                     Some(rbv::ResolvedArg::Error(guar)) => {
-                        Some(tcx.const_error_with_guaranteed(ty, guar))
+                        Some(tcx.const_error_with_guaranteed(param_ty, guar))
                     }
                     arg => bug!("unexpected bound var resolution for {:?}: {arg:?}", expr.hir_id),
                 }

--- a/compiler/rustc_trait_selection/src/traits/specialize/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/specialize/mod.rs
@@ -99,10 +99,10 @@ pub fn translate_substs<'tcx>(
             }
 
             fulfill_implication(infcx, param_env, source_trait_ref, target_impl).unwrap_or_else(
-                |_| {
+                |()| {
                     bug!(
-                        "When translating substitutions for specialization, the expected \
-                         specialization failed to hold"
+                        "When translating substitutions from {source_impl:?} to {target_impl:?}, \
+                        the expected specialization failed to hold"
                     )
                 },
             )

--- a/tests/ui/const-generics/type_mismatch.rs
+++ b/tests/ui/const-generics/type_mismatch.rs
@@ -1,5 +1,6 @@
 fn foo<const N: usize>() -> [u8; N] {
     bar::<N>() //~ ERROR mismatched types
+    //~^ ERROR the constant `N` is not of type `u8`
 }
 
 fn bar<const N: u8>() -> [u8; N] {}

--- a/tests/ui/const-generics/type_mismatch.stderr
+++ b/tests/ui/const-generics/type_mismatch.stderr
@@ -1,3 +1,15 @@
+error: the constant `N` is not of type `u8`
+  --> $DIR/type_mismatch.rs:2:5
+   |
+LL |     bar::<N>()
+   |     ^^^^^^^^
+   |
+note: required by a bound in `bar`
+  --> $DIR/type_mismatch.rs:6:8
+   |
+LL | fn bar<const N: u8>() -> [u8; N] {}
+   |        ^^^^^^^^^^^ required by this bound in `bar`
+
 error[E0308]: mismatched types
   --> $DIR/type_mismatch.rs:2:11
    |
@@ -5,7 +17,7 @@ LL |     bar::<N>()
    |           ^ expected `u8`, found `usize`
 
 error[E0308]: mismatched types
-  --> $DIR/type_mismatch.rs:5:26
+  --> $DIR/type_mismatch.rs:6:26
    |
 LL | fn bar<const N: u8>() -> [u8; N] {}
    |    ---                   ^^^^^^^ expected `[u8; N]`, found `()`
@@ -13,11 +25,11 @@ LL | fn bar<const N: u8>() -> [u8; N] {}
    |    implicitly returns `()` as its body has no tail or `return` expression
 
 error[E0308]: mismatched types
-  --> $DIR/type_mismatch.rs:5:31
+  --> $DIR/type_mismatch.rs:6:31
    |
 LL | fn bar<const N: u8>() -> [u8; N] {}
    |                               ^ expected `usize`, found `u8`
 
-error: aborting due to 3 previous errors
+error: aborting due to 4 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/const-generics/type_not_in_scope.rs
+++ b/tests/ui/const-generics/type_not_in_scope.rs
@@ -6,6 +6,5 @@ impl X {
 }
 fn getn<const N: cfg_attr>() -> [u8; N] {}
 //~^ ERROR expected type, found built-in attribute `cfg_attr`
-//~| ERROR mismatched types
 
 fn main() {}

--- a/tests/ui/const-generics/type_not_in_scope.stderr
+++ b/tests/ui/const-generics/type_not_in_scope.stderr
@@ -10,15 +10,7 @@ error[E0573]: expected type, found built-in attribute `cfg_attr`
 LL | fn getn<const N: cfg_attr>() -> [u8; N] {}
    |                  ^^^^^^^^ not a type
 
-error[E0308]: mismatched types
-  --> $DIR/type_not_in_scope.rs:7:33
-   |
-LL | fn getn<const N: cfg_attr>() -> [u8; N] {}
-   |    ----                         ^^^^^^^ expected `[u8; N]`, found `()`
-   |    |
-   |    implicitly returns `()` as its body has no tail or `return` expression
+error: aborting due to 2 previous errors
 
-error: aborting due to 3 previous errors
-
-Some errors have detailed explanations: E0308, E0412, E0573.
-For more information about an error, try `rustc --explain E0308`.
+Some errors have detailed explanations: E0412, E0573.
+For more information about an error, try `rustc --explain E0412`.

--- a/tests/ui/specialization/min_specialization/bad-const-wf-doesnt-specialize.rs
+++ b/tests/ui/specialization/min_specialization/bad-const-wf-doesnt-specialize.rs
@@ -1,0 +1,12 @@
+#![feature(min_specialization)]
+
+// An impl that has an erroneous const substitution should not specialize one
+// that is well-formed.
+
+struct S<const L: usize>;
+
+impl<const N: i32> Copy for S<N> {}
+impl<const M: usize> Copy for S<M> {}
+//~^ ERROR conflicting implementations of trait `Copy` for type `S<_>`
+
+fn main() {}

--- a/tests/ui/specialization/min_specialization/bad-const-wf-doesnt-specialize.stderr
+++ b/tests/ui/specialization/min_specialization/bad-const-wf-doesnt-specialize.stderr
@@ -1,0 +1,11 @@
+error[E0119]: conflicting implementations of trait `Copy` for type `S<_>`
+  --> $DIR/bad-const-wf-doesnt-specialize.rs:9:1
+   |
+LL | impl<const N: i32> Copy for S<N> {}
+   | -------------------------------- first implementation here
+LL | impl<const M: usize> Copy for S<M> {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `S<_>`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0119`.

--- a/tests/ui/transmutability/issue-101739-1.rs
+++ b/tests/ui/transmutability/issue-101739-1.rs
@@ -6,7 +6,7 @@ mod assert {
     pub fn is_transmutable<Src, Context, const ASSUME_ALIGNMENT: bool>()
     where
         Dst: BikeshedIntrinsicFrom<Src, Context, ASSUME_ALIGNMENT>, //~ ERROR cannot find type `Dst` in this scope
-        //~^ ERROR mismatched types
+        //~^ the constant `ASSUME_ALIGNMENT` is not of type `Assume`
     {
     }
 }

--- a/tests/ui/transmutability/issue-101739-1.stderr
+++ b/tests/ui/transmutability/issue-101739-1.stderr
@@ -4,13 +4,15 @@ error[E0412]: cannot find type `Dst` in this scope
 LL |         Dst: BikeshedIntrinsicFrom<Src, Context, ASSUME_ALIGNMENT>,
    |         ^^^ not found in this scope
 
-error[E0308]: mismatched types
-  --> $DIR/issue-101739-1.rs:8:50
+error: the constant `ASSUME_ALIGNMENT` is not of type `Assume`
+  --> $DIR/issue-101739-1.rs:8:14
    |
 LL |         Dst: BikeshedIntrinsicFrom<Src, Context, ASSUME_ALIGNMENT>,
-   |                                                  ^^^^^^^^^^^^^^^^ expected `Assume`, found `bool`
+   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: required by a bound in `BikeshedIntrinsicFrom`
+  --> $SRC_DIR/core/src/mem/transmutability.rs:LL:COL
 
 error: aborting due to 2 previous errors
 
-Some errors have detailed explanations: E0308, E0412.
-For more information about an error, try `rustc --explain E0308`.
+For more information about this error, try `rustc --explain E0412`.

